remote-config/src/fetch/fetcher.rs

```diff
@@ -0,0 +1,316 @@
+}
+
+#[derive(Default)]
+pub struct OpaqueState {
+    client_state: Vec<u8>,
+}
+
+impl<S: FileStorage> ConfigFetcher<S> {
+    pub fn new(file_storage: S, state: Arc<ConfigFetcherState<S::StoredFile>>) -> Self {
+        ConfigFetcher {
+            file_storage,
+            state,
+            timeout: AtomicU32::new(5000),
+            interval: AtomicU32::new(0),
+        }
+    }
+
+    /// Quite generic fetching implementation:
+    ///  - runs a request against the Remote Config Server,
+    ///  - validates the data,
+    ///  - removes unused files
+    ///  - checks if the files are already known,
+    ///  - stores new files,
+    ///  - returns all currently active files.
+    /// It also makes sure that old files are dropped before new files are inserted.
+    pub async fn fetch_once(
+        &mut self,
+        runtime_id: &str,
+        target: Arc<Target>,
+        config_id: &str,
+        last_error: Option<String>,
+        opaque_state: &mut OpaqueState,
+    ) -> anyhow::Result<Vec<Arc<S::StoredFile>>> {
+        let Target { service, env, app_version } = (*target).clone();
+
+        let mut cached_target_files = vec![];
+        let mut config_states = vec![];
+
+        for StoredTargetFile { state, meta, .. } in self.state.target_files_by_path.lock().unwrap().values() {
+            config_states.push(state.clone());
+            cached_target_files.push(meta.clone());
+        }
+
+        let config_req = ClientGetConfigsRequest {
+            client: Some(datadog_trace_protobuf::remoteconfig::Client {
+                state: Some(ClientState {
+                    root_version: 1,
+                    targets_version: 0,
+                    config_states,
+                    has_error: last_error.is_some(),
+                    error: last_error.unwrap_or_default(),
+                    backend_client_state: std::mem::take(&mut opaque_state.client_state),
+                }),
+                id: config_id.into(),
+                // TODO maybe not hardcode requested products?
+                products: vec!["APM_TRACING".to_string(), "LIVE_DEBUGGING".to_string()],
+                is_tracer: true,
+                client_tracer: Some(ClientTracer {
+                    runtime_id: runtime_id.to_string(),
+                    language: self.state.invariants.language.to_string(),
+                    tracer_version: self.state.invariants.tracer_version.clone(),
+                    service,
+                    extra_services: vec![],
+                    env,
+                    app_version,
+                    tags: vec![],
+                }),
+                is_agent: false,
+                client_agent: None,
+                last_seen: 0,
+                capabilities: vec![],
+            }),
+            cached_target_files,
+        };
+        let json = serde_json::to_string(&config_req)?;
+
+        // TODO: directly talking to datadog endpoint (once signatures are validated)
```

Comment:
before any of this can be done it means this is a compliant and secure rust TUF implementation, I disagree with this.

Reply:
Should not be marked todo, I agree; bad wording.